draft-ietf-quic-transport.md

```diff
@@ -2697,6 +2697,43 @@ also be forward-secure encrypted.  Since the attacker will not have the forward
 secure key, the attacker will not be able to generate forward-secure encrypted
 packets with ACK frames.
 
+## Stream fragmentation and reassembly attacks
+
+An adversarial client may attempt to
+exhaust server memory resource by performing
+a stream fragmentation and reassembly attack, similar to the UDP/ICMP
+"Teardrop" fragmentation attacks. The adversarial client would open a stream,
+and send some STREAM DATA packets containing fragments of the stream content.
+The goal of the attack is to induce the receiving implementation to commit
+memory buffers while waiting that the stream data can be reassembled.
+
+For example, a client
+might send some octets at the beginning of a
+stream and some octets much further
+away in the stream. A vulnerable server, on receiving the stream data fragments,
+might allocate memory covering the whole space from beginning to end. The client
+would then repeat the process on a large number of streams and a large number
+of connections, inducing the server to commit a large amout of memory.
+
+This attack can be mitigated by not
+committing memory for stream data reassembly,
+and simply keeping the STREAM DATA frames until enough fragments have been
+received and the data can be delivered to the application in proper sequence.
```

Comment:
I should perhaps clarify that in the above, a connection level budget is not a linear function of flow control. There is a fixed amount of internal memory, and as that is released, the congestion window is expanded. So storing lots of fragments will use memory faster and release memory slower, and thus reduce the connection level congestion window. And, when the memory fills, packets starts to drop. In this way, the worst case is that the full budget is consumed with holes, whereas a friendly peer would feel the same budget with linear stream data. The adversary can only create so many holes before the cost of whole punching is more expensive than linear data. Of course, there a endless different ways this can be handled, and it depends on the use case, risk, degradation when not under attack, etc., therefore it is hard to provide general advise.